### PR TITLE
Relax safety precondition of `Ptr::cast_unsized`

### PR DESCRIPTION
**Pull request by jswrenn**
We relax the `UnsafeCell` safety conditions on `cast_unsized` for exclusively-aliased pointers. This paves the way for removing the `NoCell` bound from `try_cast_into`, `try_cast_into_no_leftover`, and finally from `TryFromBytes::try_from_mut`.

I *think* this is sound; putting this PR up so we can discuss @joshlf.

### CODE REVIEW

File: src/pointer/ptr.rs [hunk 1]
```diff
@@ -1244,10 +1248,11 @@ mod _project {
         ///   `self` casted to a raw pointer. The pointer it returns must
         ///   reference only a subset of `self`'s bytes.
         /// - `T` is a struct or union type.
-        /// - The projected pointer contains `UnsafeCell`s at exactly the same
-        ///   ranges at which `UnsafeCell`s appear in the projected-from type.
-        ///   This is necessarily true for projections of struct fields, but not
-        ///   for all projections of union fields.
+        /// - If the aliasing of `self` is Shared, projected pointer must
```
Comment (joshlf) [inline src/pointer/ptr.rs:1251]:
```suggestion
        /// - If the aliasing of `self` is not `Exclusive`, projected pointer must
```

File: src/pointer/ptr.rs [hunk 2]
```diff
@@ -918,8 +918,9 @@ mod _casts {
         /// For all kinds of casts, the caller must promise that:
         /// - the the size of the object referenced by the resulting pointer is
         ///   less than or equal to the size of the object referenced by `self`.
-        /// - `UnsafeCell`s in `U` exist at ranges identical to those at which
-        ///   `UnsafeCell`s exist in `T`.
+        /// - if the aliasing of `self` is Shared, that `UnsafeCell`s in `U`
```
Comment (joshlf) [inline src/pointer/ptr.rs:921]:
```suggestion
        /// - if the aliasing of `self` is not `Exclusive`, that `UnsafeCell`s in `U`
```

File: src/pointer/ptr.rs [hunk 3]
```diff
@@ -983,7 +984,10 @@ mod _casts {
             //    store this memory region treating `UnsafeCell`s as existing at
             //    different ranges than they exist in `U`. This is true by
             //    invariant on Ptr<'a, T, I>, and preserved through the cast to
-            //    `U` by contract on the caller.
+            //    `U` by contract on the caller:
+            //     - If `ptr` is exclusively aliased, no other references exist.
```
Comment (joshlf) [inline src/pointer/ptr.rs:988]:
```suggestion
            //     - If `ptr` is exclusively aliased, no other live references exist.
```

File: src/pointer/ptr.rs [hunk 4]
```diff
@@ -63,9 +63,9 @@ mod def {
         /// 10. During the lifetime 'a, no code will load or store this memory
         ///     region treating `UnsafeCell`s as existing at different ranges
         ///     than they exist in `T`.
-        /// 11. During the lifetime 'a, no reference will exist to this memory
-        ///     which treats `UnsafeCell`s as existing at different ranges than
-        ///     they exist in `T`.
+        /// 11. During the lifetime 'a, no live reference will exist to this
```
Comment (joshlf) [inline src/pointer/ptr.rs:66]:
```suggestion
        /// 11. During the lifetime 'a, no live reference and no live `Ptr` will exist to this
```

**Review by joshlf — CHANGES_REQUESTED** [review]

File: src/pointer/ptr.rs (hunk 1 above)
Comment (jswrenn) [reply src/pointer/ptr.rs]: Done

File: src/pointer/ptr.rs (hunk 3 above)
Comment (jswrenn) [reply src/pointer/ptr.rs]: Done

File: src/pointer/ptr.rs (hunk 2 above)
Comment (jswrenn) [reply src/pointer/ptr.rs]: Done

File: src/pointer/ptr.rs (hunk 4 above)
Comment (jswrenn) [reply src/pointer/ptr.rs]: Done

**Review by joshlf — CHANGES_REQUESTED** [review]
Anywhere that we reason about `UnsafeCell` non-overlap being okay in virtue of exclusive aliasing, we should add:

```rust
// TODO(#896), TODO(https://github.com/rust-lang/unsafe-code-guidelines/issues/495): Blah
// blah blah before the next stable release.
```

**Comment by jswrenn** [comment]
Superseded by #1211.

**Pull request closed without merging**